numpy/core/src/multiarray/multiarraymodule.c

```diff
@@ -1045,6 +1045,20 @@ PyArray_MatrixProduct2(PyObject *op1, PyObject *op2, PyArrayObject* out)
         result = (PyArray_NDIM(ap1) == 0 ? ap1 : ap2);
         result = (PyArrayObject *)Py_TYPE(result)->tp_as_number->nb_multiply(
                                         (PyObject *)ap1, (PyObject *)ap2);
+        if (out) {
+            if (PyArray_AssignArray(out, (PyArrayObject *)result,
+                        NULL, NPY_DEFAULT_ASSIGN_CASTING) < 0) {
+                Py_DECREF(result);
+                Py_DECREF(ap1);
+                Py_DECREF(ap2);
+                return NULL;
+            }
+            Py_DECREF(result);
+            Py_DECREF(ap1);
+            Py_DECREF(ap2);
+            Py_INCREF(out);
+            return (PyObject *)out;
+        }
```

Comment:
Thanks!  The file already includes `"number.h"` and that makes calling the ufunc directly easy (and the ufunc supports `out`).

So I think we can simplify this a fair bit by doing something like (i.e. calling the ufunc directly):
```
result = PyObject_CallFunctionObjArgs(n_ops.multiply, ap1, ap2, out, NULL);
/* cleanup */
return result;
```
(that includes removint that weird `result` setup.

Reply:
Hi! Thank you for the feedback.
I've simplified the code with your suggestion and everything seems OK.